File: Problems/LongestIncreasingSubsequence.java
/**
 * Given an integer array nums, return the length of the longest strictly increasing subsequence.
 * 
 * A subsequence is a sequence that can be derived from an array 
 * by deleting some or no elements without changing the order of the remaining elements. 
 * For example, [3,6,2,7] is a subsequence of the array [0,3,1,6,2,2,7].
 */
package Problems;

import java.util.*;

public class LongestIncreasingSubsequence {
    // O(nlogn) time complexity
    public static int lengthOfLIS(int[] nums) {
        int res = 1;
        if (nums.length == 1)
            return res;
        List<Integer> subArray = new LinkedList<>();
        subArray.add(nums[0]);
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] < subArray.get(subArray.size() - 1))
                helper(nums[i], subArray);
            else if (nums[i] > subArray.get(subArray.size() - 1))
                subArray.add(nums[i]);
        }
        return subArray.size();
    }

    private static void helper(int value, List<Integer> subArray) {
        for (int i = 0; i < subArray.size(); i++) {
            if (value <= subArray.get(i)) {
                subArray.set(i, value);
                return;
            }
        }
    }
    
    public static void main(String[] args) {
        int[] nums = { 4, 10, 4, 3, 8, 9 };
        System.out.println(lengthOfLIS(nums));
    }
}

File: Problems/LongestIncreasingSubsequence.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
   public:
    // build subsequence with binary search
    // time complexity: O(nlogn)
    // space complexity: O(n)
    void replace(vector<int>& sub, int num) {
        if (sub[0] > num) {
            sub[0] = num;
            return;
        }

        int start = 0;
        int end = sub.size() - 1;
        while (start <= end) {
            int mid = (end - start) / 2 + start;
            if (sub[mid] == num || (sub[mid] > num && sub[mid - 1] < num)) {
                sub[mid] = num;
                return;
            }
            if (sub[mid] > num) {
                end = mid;
            } else if (sub[mid] < num) {
                start = mid + 1;
            }
        }
    }

    int lengthOfLIS(vector<int>& nums) {
        vector<int> sub;
        for (auto num : nums) {
            if (sub.size() == 0 || num > sub[sub.size() - 1])
                sub.push_back(num);
            else if (sub[sub.size() - 1] == num)
                continue;
            else if (num < sub[sub.size() - 1])
                replace(sub, num);
        }
        return sub.size();
    }
};